instrumentation/netty-codec-http/src/main/java/brave/netty/http/TracingHttpServerHandler.java

```diff
@@ -0,0 +1,99 @@
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import java.net.InetSocketAddress;
+import zipkin2.Endpoint;
+
+final class TracingHttpServerHandler extends ChannelDuplexHandler {
+  final Tracer tracer;
+  final HttpNettyAdapter adapter;
+  final TraceContext.Extractor<HttpHeaders> extractor;
+  final HttpSampler sampler;
+  final HttpServerParser parser;
+
+  TracingHttpServerHandler(HttpTracing httpTracing) {
+    tracer = httpTracing.tracing().tracer();
+    sampler = httpTracing.serverSampler();
+    parser = httpTracing.serverParser();
+    adapter = new HttpNettyAdapter();
+    extractor = httpTracing.tracing().propagation().extractor(HttpHeaders::get);
+  }
+
+  @Override public void channelRead(ChannelHandlerContext ctx, Object msg) {
+    if (!(msg instanceof HttpRequest)) {
+      ctx.fireChannelRead(msg); // superclass does not throw
+      return;
+    }
+
+    HttpRequest request = (HttpRequest) msg;
+
+    Span span = nextSpan(extractor.extract(request.headers()), request).kind(Span.Kind.SERVER);
+    ctx.channel().attr(NettyHttpTracing.SPAN_ATTRIBUTE).set(span);
+
+    // Place the span in scope so that downstream code can read trace IDs
+    try (SpanInScope ws = tracer.withSpanInScope(span)) {
+      if (!span.isNoop()) {
+        parser.request(adapter, request, span);
+        maybeParseClientAddress(ctx.channel(), request, span);
+        span.start();
+      }
+      ctx.fireChannelRead(msg);
+    }
+  }
+
+  /** Like {@link HttpServerHandler}, but accepts a channel */
+  void maybeParseClientAddress(Channel channel, HttpRequest request, Span span) {
+    Endpoint.Builder remoteEndpoint = Endpoint.newBuilder();
+    if (adapter.parseClientAddress(request, remoteEndpoint)) {
+      span.remoteEndpoint(remoteEndpoint.build());
+    } else {
+      InetSocketAddress remoteAddress = (InetSocketAddress) channel.remoteAddress();
+      span.remoteEndpoint(Endpoint.newBuilder()
+          .ip(remoteAddress.getAddress())
+          .port(remoteAddress.getPort())
+          .build());
+    }
+  }
+
+  /** Creates a potentially noop span representing this request */
+  // copy/pasted from HttpServerHandler.nextSpan
+  Span nextSpan(TraceContextOrSamplingFlags extracted, HttpRequest request) {
+    if (extracted.sampled() == null) { // Otherwise, try to make a new decision
+      extracted = extracted.sampled(sampler.trySample(adapter, request));
+    }
+    return extracted.context() != null
+        ? tracer.joinSpan(extracted.context())
+        : tracer.nextSpan(extracted);
+  }
+
+  @Override public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise prm) {
+    Span span = ctx.channel().attr(NettyHttpTracing.SPAN_ATTRIBUTE).get();
+    if (span == null || !(msg instanceof HttpResponse)) {
+      ctx.write(msg, prm);
+      return;
+    }
+
+    // Place the span in scope so that downstream code can read trace IDs
+    HttpResponse response = (HttpResponse) msg;
```

Comment:
Just a quick note - depending on how the Netty pipeline is setup the `HttpResponse` message may or may not represent a full response. For example you could send the response in chunks where it starts with a `HttpResponse`, continues with any number of `HttpContent`s, and ends with a `LastHttpContent`. For responses that have really big payloads (and/or slow connections) this means the span could technically complete a while before the response is fully sent. 

Whether this matters or not is definitely debatable - I could even see arguments that completing the span at this point is desirable behavior. Plenty of stacks don't even give you an ability to know when the response has been fully sent, so it's not like this is much different than other instrumentations. 

Reworking things so that the span completes after the last chunk is sent would probably involve keeping track of the `HttpResponse` since it looks like you're using that for the `parser`, combined with waiting for the `LastHttpContent` message and attaching a listener that triggers off the `ctx.write(...)` call when it's done sending. That's probably enough extra complexity that it's not worth the debatable benefit - just thought I'd mention it so it can be a conscious decision.